model/flow/header.go

```diff
@@ -72,6 +73,20 @@ func (h Header) Fingerprint() []byte {
 	return fingerprint.Fingerprint(h.Body())
 }
 
+var previd_hdr Identifier
+var prev_hdr Header
```

Comment:
Could you please explain why this is safe under concurrent operations. 
IDs for headers are computed by many threads concurrently; you cache the last header and its ID value in a globally shared variable, without any synchronization. 

I think data races will make the check fails often, i.e. we are wasting performance. But, even worse, I think data races could lead to wrong IDs being reported:
* let `h` be some header; consider the execution of `h.ID()`
* assume that this check finds that `h` is identical to `prev_hdr`: https://github.com/onflow/flow-go/blob/41b3fc4ac3a3f9e38b805da3b21164ed4cef7d79/model/flow/header.go#L131-L132
Even you find the current header is equal 
   * now we return `previd_hdr`, but this value could have been just overwritten but a different thread, causing us to return the wrong ID. 

A similar bug could happen between these three operations, as they are not atomic either: https://github.com/onflow/flow-go/blob/41b3fc4ac3a3f9e38b805da3b21164ed4cef7d79/model/flow/header.go#L138-L143

Reply:
I wasn't seeing evidence of multi-threading during testing, but you're right, it could happen.  For instance, the logging would have been jumbled if it were multi-threaded, but it wasn't.
I'll put something around it, just in case.